### PR TITLE
Quick doc on how to add a new type to SerDe

### PR DESCRIPTION
**Pull request by GoEddie**
Doc on how to add a new type to SerDe, hope it helps someone else

### CODE REVIEW

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+the code that will perform the write you need to add a case to `GetTypeId` for your new identifier:
+
+```csharp
+
+    case TypeCode.Double:
+        return s_doubleTypeId;
+
+```
+
+The next step is to add the derserialization code to scala (read code).
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
## 2. Derserialize this new type in JVM with read method

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      case 'l' => readList(dis)
+      case 'D' => readDate(dis)
+      case 't' => readTime(dis)
+      case 'j' => JVMObjectTracker.getObject(readString(dis))
+      case 'R' => readRowArr(dis)
+      case _ => throw new IllegalArgumentException(s"Invalid type $dataType")
+    }
+  }
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you please show one example to be consistent with E2E example? e.g. `DoubleArrayArray`.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    val len = readInt(in)
+    Row.fromSeq((0 until len).map(_ => readObject(in)))
+  }
+
+  def readBytesArr(in: DataInputStream): Array[Array[Byte]] = {
+    val len = readInt(in)
+    (0 until len).map(_ => readBytes(in)).toArray
+  }
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you please show one example to be consistent with E2E example? e.g. `DoubleArrayArray`.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    case TypeCode.Double:
+        return s_doubleTypeId;
+
+```
+
+The next step is to add the derserialization code to scala (read code).
+
+In src/main/scala/org/apache/spark/api/dotnet/SerDe.scala (These are versioned so you may need to 
+try and find it). In the method `readTypedObject` add a new case statement for your new type 
+identifier:
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Please write this in one line.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+```csharp
+
+    case TypeCode.Double:
+        return s_doubleTypeId;
+
+```
+
+The next step is to add the derserialization code to scala (read code).
+
+In src/main/scala/org/apache/spark/api/dotnet/SerDe.scala (These are versioned so you may need to
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
In [SerDe.scala](https://github.com/dotnet/spark/blob/master/src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/SerDe.scala) (This need to be changed in all Spark version 2.3.x, 2.4.x and 3.0.x). In the method `readTypedObject` add a new case statement for your new type 
identifier:
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  def readBytesArr(in: DataInputStream): Array[Array[Byte]] = {
+    val len = readInt(in)
+    (0 until len).map(_ => readBytes(in)).toArray
+  }
+
+```
+
+That should be everything you need to be able to send a new type from the CLR to the JVM but we now
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
That should be everything you need to be able to send a new type from the CLR to the JVM.
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+ need to write back from the JVM to the CLR.
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion

## 3. Serialize this new type in JVM with write method

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    out.writeDouble(value.getTime.toDouble / 1000.0)
+  }
+
+ def writeDoubleArr(out: DataOutputStream, value: Array[Double]): Unit = {
+    writeType(out, "double")
+    out.writeInt(value.length)
+    value.foreach(v => out.writeDouble(v))
+  }
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you please show one example to be consistent with E2E example? e.g. `DoubleArrayArray`.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+The final step is to be able to read the value on the CLR side that the JVM side has written. Back 
+in csharp find csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs add your type identifier to 
+`CallJavaMethod` in the switch statement `switch (typeAsChar) //TODO: Add support for other types.`
+you will likely find that the type you want to implement should be implemented inside 
+`ReadCollection`.
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Please write these in one line.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+ def writeDoubleArr(out: DataOutputStream, value: Array[Double]): Unit = {
+    writeType(out, "double")
+    out.writeInt(value.length)
+    value.foreach(v => out.writeDouble(v))
+  }
+
+```
+
+The final step is to be able to read the value on the CLR side that the JVM side has written. Back
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
## 4. Add this new type in [JvmBridge.cs](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs#L151) to read the value on CLR side
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+ def writeDoubleArr(out: DataOutputStream, value: Array[Double]): Unit = {
+    writeType(out, "double")
+    out.writeInt(value.length)
+    value.foreach(v => out.writeDouble(v))
+  }
+
+```
+
+The final step is to be able to read the value on the CLR side that the JVM side has written. Back 
+in csharp find csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs add your type identifier to
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion

The final step is to be able to read the value on the CLR side that the JVM side has written. Back in csharp find csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs add your type identifier to `CallJavaMethod` in the switch statement `switch (typeAsChar) //TODO: Add support for other types.` You will likely find that the type you want to implement should be implemented inside `ReadCollection`.
```

**Comment by GoEddie** [comment]
Thanks for the review @elvaliuliuliu I think I have made the changes you suggested, let me know if you think anything else should change.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            SerDe.Write(destination, doubleArray.Length);
+            foreach (double d in doubleArray)
+            {
+                SerDe.Write(destination, d);
+            }
+        }
+        break;
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you remove the empty lines and adjust the indentation?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    if (type == typeof(int[]) ||
+        type == typeof(long[]) ||
+        type == typeof(double[]) ||
+        type == typeof(double[][]) ||
+        typeof(IEnumerable<byte[]>).IsAssignableFrom(type) ||
+        typeof(IEnumerable<string>).IsAssignableFrom(type))
+    {
+        return s_arrayTypeId;
+    }
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you remove the empty lines and adjust the indentation?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/PayloadHelper.cs#L20
+
+Example:
+
+```csharp
+
+    private static readonly byte[] s_doubleArrayArrayTypeId = new[] { ( byte)'A' };z
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you remove the empty lines and adjust the indentation? Also please remove `z` in the end.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+            {
+                SerDe.Write(destination, d);
+            }
+        }
+        break;
+
+```
+
+Typically the type identifier is written, then the length and then the data. Once you have written 
+the code that will perform the write you need to add a case to `GetTypeId` for your new identifier:
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you merge these into one line?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+In [SerDe.scala](https://github.com/dotnet/spark/blob/master/src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/SerDe.scala) (This need to be changed in all Spark version 2.3.x, 2.4.x and 3.0.x). 
+
+In either the method `readTypedObject` or `readList` add a new case statement for your new type 
+identifier:
+
+```scala
+
+  case 'A' => readDoubleArrArr(dis)
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you remove the empty lines and adjust the indentation?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+Once you have read the type, you can implement your method to read the data you wrote in csharp, examples:
+
+```scala
+
+   def readDoubleArrArr(in: DataInputStream): Array[Array[Double]] = {
+    val len = readInt(in)
+    (0 until len).map(_ => readDoubleArr(in)).toArray
+  }
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you remove the empty lines and adjust the indentation?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+That should be everything you need to be able to send a new type from the CLR to the JVM.
+
+### 3. Serialize this new type in JVM with write method
+
+In src/main/scala/org/apache/spark/api/dotnet/SerDe.scala the function `writeObject` contains a switch for each type, a couple of examples:
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
In [SerDe.scala](https://github.com/dotnet/spark/blob/master/src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/SerDe.scala) (This need to be changed in all Spark version 2.3.x, 2.4.x and 3.0.x), the function `writeObject` contains a switch for each type, a couple of examples:
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+```scala
+
+ case "[[D" =>
+  writeType(dos, "list")
+  writeDoubleArrArr(dos, value.asInstanceOf[Array[Array[Double]]])
+
+```
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Can you remove the empty lines and adjust the indentation?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+### 4. Add this new type in [JvmBridge.cs](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs#L151) to read the value on CLR side
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
### 4. Add this new type in [JvmBridge.cs](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs) to read the value on CLR side
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+```
+
+That should be everything you need to be able to send a new type from the CLR to the JVM.
+
+### 3. Serialize this new type in JVM with write method
+
+In src/main/scala/org/apache/spark/api/dotnet/SerDe.scala the function `writeObject` contains a switch for each type, a couple of examples:
+
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: How about adding `case "doublearray" => dos.writeByte('A')` also in this section?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+```scala
+
+ case "[[D" =>
+  writeType(dos, "list")
+  writeDoubleArrArr(dos, value.asInstanceOf[Array[Array[Double]]])
+
+```
+
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: How about adding the following write function also?
```
def writeDoubleArrArr(out: DataOutputStream, value: Array[Array[Double]]): Unit = {
    writeType(out, "doublearray")
    out.writeInt(value.length)
    value.foreach(v => writeDoubleArr(out, v))
}
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+### 4. Add this new type in [JvmBridge.cs](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs#L151) to read the value on CLR side
+
+The final step is to be able to read the value on the CLR side that the JVM side has written. Back in csharp find csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs add your type identifier to `CallJavaMethod` in the switch statement `switch (typeAsChar) //TODO: Add support for other types.` You will likely find that the type you want to implement should be implemented inside `ReadCollection`.
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Let's add the [sample code](https://github.com/GoEddie/spark/blob/1968f59c22df9808ab210fa7bf94ab5cdd48a9f2/src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs#L368) here as an example.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+        return s_arrayTypeId;
+    }
+```
+
+### 2. Derserialize this new type in JVM with read method
+
+In [SerDe.scala](https://github.com/dotnet/spark/blob/master/src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/SerDe.scala) (This need to be changed in all Spark version 2.3.x, 2.4.x and 3.0.x). 
+
+In either the method `readTypedObject` or `readList` add a new case statement for your new type 
+identifier:
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]: Let's merge these into one line together.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+
+This how-to provides general instructions on how to add new types to serialization and deserialization between CLR (Common Language Runtime) and JVM (Java virtual machine). The IPC (Inter-process communication) mechanism between the CLR and the JVM uses serialization to communicate. Each type has a type identifier and then an optional length followed by the actual data.
+
+Not every type has been implemented so if you need to use a type that isn't supported and you get the error "Type {0} not supported yet", you will need to implement the type.
+
+## Steps to follow
+
+### 1. Choose a new type identifier and add it to:
+
+https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/PayloadHelper.cs#L20
```
[the remark this answers is not in the document]
Comment (GoEddie) [reply docs/coding-guidelines/add-types-to-SerDe.md]: removed the url (think you mean that and not the ### 1. line

**Comment by GoEddie** [comment]
@elvaliuliuliu I have updated the doc, let me know if you think there should be more changes :)

File: docs/coding-guidelines/add-types-to-SerDe.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (elvaliuliuliu) [reply docs/coding-guidelines/add-types-to-SerDe.md]: That's correct. I meant to remove the url. Thanks!

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+
+## Steps to follow
+Please see the following E2E example of how the new type `DoubleArrayArray` is added to serialization and deserialization between CLR and JVM.
+
+### 1. Choose a new type identifier and add it in [PayloadHelper.cs](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/PayloadHelper.cs) on CLR side
+
+Example:
+
+```csharp
+private static readonly byte[] s_doubleArrayArrayTypeId = new[] { ( byte)'A' };z
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
private static readonly byte[] s_doubleArrayArrayTypeId = new[] { ( byte)'A' };
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+    writeType(out, "doublearray")
+    out.writeInt(value.length)
+    value.foreach(v => writeDoubleArr(out, v))
+}
+```
+
+`WriteType` also needs to know how to write the specific type identifier:
+
+```scala
+case "doublearray" => dos.writeByte('A'
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
case "doublearray" => dos.writeByte('A')
```

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+
+`WriteType` also needs to know how to write the specific type identifier:
+
+```scala
+case "doublearray" => dos.writeByte('A')
+```
+
+### 4. Add this new type in [JvmBridge.cs](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs) to read the value on CLR side
+
+The final step is to be able to read the value on the CLR side that the JVM side has written. Back in csharp find csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs add your type identifier to `CallJavaMethod` in the switch statement `switch (typeAsChar) //TODO: Add support for other types.` You will likely find that the type you want to implement should be implemented inside `ReadCollection`.
```
Comment (elvaliuliuliu) [inline docs/coding-guidelines/add-types-to-SerDe.md]:
```suggestion
The final step is to be able to read the value on the CLR side that the JVM side has written. Add your type identifier to `CallJavaMethod` in the switch statement `switch (typeAsChar) //TODO: Add support for other types.` in [JvmBridge.cs](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Interop/Ipc/JvmBridge.cs). You will likely find that the type you want to implement should be implemented inside `ReadCollection`.

Add new type identifier for `DoubleArrayArray` as below:
```

**Comment by imback82** [comment]
Oh I noticed that the title has changed to WIP. Do you want me to wait on this?

**Comment by GoEddie** [comment]
No it is ok now, I added WIP as when I have been merging on github the changes the file was going a bit screwy so checking that it was all ok.

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+}
+```
+
+### 2. Derserialize this new type in JVM with read method
+
+In [SerDe.scala](https://github.com/dotnet/spark/blob/master/src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/SerDe.scala) (This need to be changed in all Spark version 2.3.x, 2.4.x and 3.0.x). 
+
+In either the method `readTypedObject` or `readList` add a new case statement for your new type 
+identifier:
+
```
[the remark this answers is not in the document]
Comment (elvaliuliuliu) [reply docs/coding-guidelines/add-types-to-SerDe.md]: Can we remove these redundant sentences before merging it? Thanks!

[the remark this answers is not in the document]
Comment (imback82) [reply docs/coding-guidelines/add-types-to-SerDe.md]: @GoEddie can you address this comment?

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+    typeof(IEnumerable<byte[]>).IsAssignableFrom(type) ||
+    typeof(IEnumerable<string>).IsAssignableFrom(type))
+{
+    return s_arrayTypeId;
+}
+```
+
+### 2. Derserialize this new type in JVM with read method
+
+In [SerDe.scala](https://github.com/dotnet/spark/blob/master/src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/SerDe.scala) (This need to be changed in all Spark version 2.3.x, 2.4.x and 3.0.x), add a new case statement for your new type identifier in either `readTypedObject` or `readList` method depending on the new type:
```
Comment (imback82) [inline docs/coding-guidelines/add-types-to-SerDe.md]: nit: This needs

File: docs/coding-guidelines/add-types-to-SerDe.md [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+  val len = readInt(in)
+  (0 until len).map(_ => readDoubleArr(in)).toArray
+}
+```
+
+That should be everything you need to be able to send a new type from the CLR to the JVM.
+
+### 3. Serialize this new type in JVM with write method
+
+In [SerDe.scala](https://github.com/dotnet/spark/blob/master/src/scala/microsoft-spark-2.3.x/src/main/scala/org/apache/spark/api/dotnet/SerDe.scala) (This need to be changed in all Spark version 2.3.x, 2.4.x and 3.0.x), the function `writeObject` contains a switch for each type:
```
Comment (imback82) [inline docs/coding-guidelines/add-types-to-SerDe.md]: nit: This needs

**Comment by GoEddie** [comment]
@imback82 i've removed the duplicate comment, sorry I had forgotten about this one!

**Comment by imback82** [comment]
@GoEddie For some reason, I come up as a committer for this PR, which is not true. I will revert this change and can you create a new PR so you get the right credit? Sorry about that!

**Pull request merged**